File: app/src/main/java/com/egecius/rxjava2_demo_2/rx/MapExamples/FlatMapExamples.java
package com.egecius.rxjava2_demo_2.rx.MapExamples;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.Single;
import io.reactivex.SingleSource;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Function;

public class FlatMapExamples {

	/** Takes list & slices it into its elements */
	public Observable<String> flatmap(final List<String> list) {
		return Observable.just(list)
				.flatMap(new Function<List<String>, ObservableSource<String>>() {
					@Override
					public ObservableSource<String> apply(@NonNull final List<String> strings)
							throws
							Exception {
						return Observable.fromIterable(strings);
					}
				});
	}

	/** Takes list & slices it into its elements. Same effect as using flatmap above */
	public Observable<String> fromIterable(final List<String> list) {
		return Observable.fromIterable(list);
	}

	public Single<String> flatmapOnSingle(final Integer integerOuter) {
		return Single.just(integerOuter)
				.flatMap(new Function<Integer, SingleSource<String>>() {
					@Override
					public SingleSource<String> apply(@NonNull final Integer integerInner) throws
							Exception {
						return Single.just(String.valueOf(integerInner));
					}
				});
	}

	public Observable<String> flatMapOneToOne(final Integer integer) {
		return Observable.just(integer)
				.flatMap(new Function<Integer, Observable<String>>() {
					@Override
					public Observable<String> apply(@NonNull final Integer integer) throws
							Exception {
						return Observable.just(String.valueOf(integer));
					}
				});
	}
}

File: app/src/main/java/com/egecius/rxjava2_demo_2/rx/map/FlatMapExamples.kt
package com.egecius.rxjava2_demo_2.rx.map

import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.functions.Function

class FlatMapExamples {

    /** Takes list & slices it into its elements  */
    fun flatMap(list: List<String>): Observable<String> {
        return Observable.just(list)
                .flatMap { strings -> Observable.fromIterable(strings) }
    }

    /** Takes list & slices it into its elements. Same effect as using [.flatMap]}  */
    fun fromIterable(list: List<String>): Observable<String> {
        return Observable.fromIterable(list)
    }

    fun flatMapOnSingle(integerOuter: Int): Single<String> {
        return Single.just(integerOuter)
                .flatMap { integerInner -> Single.just(integerInner.toString()) }
    }

    fun flatMapOneToOne(integer: Int): Observable<String> {
        return Observable.just(integer)
                .flatMap { Observable.just(it.toString()) }
    }

    fun flatMapOneToNone(integer: Int): Observable<String> {
        return Observable.just(integer)
                .flatMap { Observable.empty<String>() }
    }

    fun flatMapOneToSometimes(list: List<Int>): Observable<Int> {
        return Observable.fromIterable(list)
                .flatMap(object : Function<Int, Observable<Int>> {
                    override fun apply(integer: Int): Observable<Int> {

                        return if (isEven(integer)) {
                            Observable.empty()
                        } else {
                            Observable.just(integer)
                        }
                    }

                    private fun isEven(integer: Int?): Boolean {
                        return integer!! % 2 == 0
                    }
                })
    }

    fun singleFlatMapObservable(list: List<Int>): Observable<Int> {
        return Single.just(list)
                .flatMapObservable { integers -> Observable.fromIterable(integers) }
    }
}
